Source/UI/PaymentForm/View/ExpiryDate/ExpiryDateFormatter.swift

```diff
@@ -0,0 +1,114 @@
+//
+//  ExpiryDateFormatter.swift
+//
+//
+//  Created by Alex Ioja-Yang on 16/05/2023.
+//
+import Foundation
+import Checkout
+
+struct ExpiryDateFormatter {
+
+    let separator: String
+    let dateFormatTextCount: Int
+    private let cardValidator: CardValidating
+
+    init(componentSeparator: String = "/",
+         cardValidator: CardValidating = CardValidator(environment: .sandbox)) {
+        self.separator = componentSeparator
+        self.dateFormatTextCount = 2 + componentSeparator.count + 2
+        self.cardValidator = cardValidator
+    }
+
+    func createCardExpiry(from input: String) -> Result<ExpiryDate, ExpiryDateError> {
+        let components = getComponents(from: input)
+        guard let month = components.month,
+              let year = components.year else {
+            return .failure(.invalidCode)
+        }
+
+        let validationOutcome = cardValidator.validate(expiryMonth: month, expiryYear: year)
+        switch validationOutcome {
+        case .success(let expiryDate):
+            return .success(expiryDate)
+        case .failure:
+            return .failure(.invalidCode)
+        }
+    }
+
+    /// Propose a display formatted expiry date that may be shown to the user
+    /// If the cardScheme is `unknown`, this validates that the cvv is conforming to internal generic standards
+    /// - Parameters:
+    ///   - input: The input as received from the user facing UI component
+    /// - Returns: Tuple made of a formatted input for displaying to the user and an error if presented input is not valid.
+    /// If displayString is nil, the text in the UI component should be rejected at last valid input should be maintained!
+    func formatForDisplay(input: String) -> (displayString: String?, validationError: ValidationError.ExpiryDate?) {
```

Comment:
@alex-ioja-yang-cko Has all of this logic been reused or added newly? I'm just asking so we can perform testing as required before release.

Reply:
I've actually followed TDD whilst building this component, with tests being written before I even wrote the logic for how to solve it. On completion of the original version, as I integrated it and eventually manually tested, found less obvious edge cases which have also had additional unit tests added. So overall I am pretty confident it's a solid helper object for our UI. Full commit history is [on this branch](https://github.com/checkout/frames-ios/compare/release/4.0.5...alex/pimob-1883-refactor-uitest), but due to the noise in refactoring to even enable understanding the problem, I have copied the final version from those objects (hence fewer and cleaner commit history here - @okhan-okbay-cko should be happy with this 😬)

Reply:
But yes, it would be good to do some basic regression testing before an actual release. I am just reassured by the fact my keyboard has tried inputting more varied content and copied/pasted in formats that I do not know if you're even able to reproduce when smoke testing 😂